relengapi/blueprints/slaveloan/__init__.py

```diff
@@ -0,0 +1,233 @@
+@p.slaveloan.admin.require()
+@apimethod(None, body=rest.LoanAdminRequest)
+def new_loan_from_admin(body):
+    "Creates a new loan entry"
+    if not body.status:
+        raise BadRequest("Missing Status Field")
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.bugzilla_email:
+        raise BadRequest("Missing Bugzilla E-Mail")
+    if body.status != 'PENDING':
+        if not body.fqdn:
+            raise BadRequest("Missing Machine FQDN")
+        if not body.ipaddr:
+            raise BadRequest("Missing Machine IP Address")
+
+    session = g.db.session('relengapi')
+    try:
+        if body.status != 'PENDING':
+            m = Machines.as_unique(session, fqdn=body.fqdn,
+                                   ipaddr=body.ipaddr)
+        h = Humans.as_unique(session, ldap=body.LDAP,
+                             bugzilla=body.bugzilla)
+    except sa.exc.IntegrityError:
+        raise InternalServerError("Integrity Error from Database, please retry.")
+
+    if body.status != 'PENDING':
+        l = Loans(status=body.status, human=h, machine=m)
+    else:
+        l = Loans(status=body.status, human=h)
+    history = History(for_loan=l,
+                      timestamp=tz.utcnow(),
+                      msg="Adding to slave loan tool via admin interface")
+    session.add(l)
+    session.add(history)
+    session.commit()
+    return None  # ?rest.WSME_New_Loan({'loan': l.to_wsme()})
+
+
+@bp.route('/loans/request', methods=['POST'])
+@apimethod(None, body=rest.LoanRequest)
+def new_loan_request(body):
+    "User Loan Requesting, returns the id of the loan"
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.requested_slavetype:
+        raise BadRequest("Missing slavetype")
+
+    slavetype = slave_to_slavetype(body.requested_slavetype)
+    if not slavetype:
+        raise BadRequest("Unsupported slavetype")
+
+    if not body.bugzilla_email:
+        # Set bugzilla e-mail to ldap e-mail by default
+        body.bugzilla_email = body.ldap_email
+
+    session = g.db.session('relengapi')
+    try:
+        h = Humans.as_unique(session, ldap=body.ldap_email,
+                             bugzilla=body.bugzilla_email)
+    except sa.exc.IntegrityError:
+        raise InternalServerError("Integrity Error from Database, please retry.")
+
+    if body.loan_bug_id:
+        l = Loans(status="PENDING", human=h, bug_id=body.loan_bug_id)
+    else:
+        l = Loans(status="PENDING", human=h)
+
+    history = History(for_loan=l,
+                      timestamp=tz.utcnow(),
+                      msg="Requesting loan for slavetype %s (original: '%s')" %
+                          (slavetype, body.requested_slavetype))
+    session.add(l)
+    session.add(history)
+    session.commit()
+    chain_of_stuff = task_groups.generate_loan(loanid=l.id, slavetype=slavetype)
+    chain_of_stuff.delay()
+    return None  # ?rest.WSME_New_Loan({'loan': l.to_wsme()})
```

Comment:
Opted for "return a Loan" (since its still a valid object without a machine). :white_check_mark: